### DISCUSSION TITLE
Why do instances that have been stopped automatically start after a period of time? How can I change this?

### DISCUSSION THREAD

**Discussion by wbadyx** [awslabs/scale-out-computing-on-aws#131] [Q&A] [answered]
Why do instances that have been stopped automatically start after a period of time? How can I change this?

**Comment by mcrozes** [comment, marked as answer]
Hello @wbadyx  -

I assume you are referring to the Virtual Desktop instances (linux or windows)? This is possible you have a scheduler configured (default to 9-5), and if you manually stop your desktop within this timeframe, SOCA will automatically restart it (as it honors VDI scheduling). 

You can learn more on https://awslabs.github.io/scale-out-computing-on-aws/web-interface/create-virtual-desktops/#configure-your-desktop-auto-startstop-time

**Reply by wbadyx** [reply]
Got it, thank you for your reply.

**Reply by NevilleGit** [reply]
> Hello @wbadyx -
> 
> I assume you are referring to the Virtual Desktop instances (linux or windows)? This is possible you have a scheduler configured (default to 9-5), and if you manually stop your desktop within this timeframe, SOCA will automatically restart it (as it honors VDI scheduling).
> 
> You can learn more on https://awslabs.github.io/scale-out-computing-on-aws/web-interface/create-virtual-desktops/#configure-your-desktop-auto-startstop-time

Hello @mcrozes ,

As my understanding, the Virtual Desktop instances launched with a scheduler configured (default to 9-5), is that possible to modify the default setting from "custom schedule(9-5)" to "stopped all day" ?

**Reply by jsynack** [reply]
Greetings @NevilleGit , 
> As my understanding, the Virtual Desktop instances launched with a scheduler configured (default to 9-5), is that possible to modify the default setting from "custom schedule(9-5)" to "stopped all day" ?

If you would like to adjust the defaults - you can update the code block here: https://github.com/awslabs/scale-out-computing-on-aws/blob/main/source/soca/cluster_web_ui/config.py#L137-L166

Note there is a difference for the Linux and Windows DCV hosts.

If the global changes do not give you enough granularity to specify the schedules that work for you - you can also update the sections here:

Windows:  https://github.com/awslabs/scale-out-computing-on-aws/blob/main/source/soca/cluster_web_ui/api/v1/dcv/create_windows_desktop.py#L505-L546

Linux: https://github.com/awslabs/scale-out-computing-on-aws/blob/main/source/soca/cluster_web_ui/api/v1/dcv/create_linux_desktop.py#L664-L705

The values are the start/stop times (expressed as minutes of the 24-hour day). 

So a `start` and `stop` of `0` would be `stopped all day` as an example.

NOTE: This area is slated for quite a bit of rewrite.   I would be interested to get an idea from you what you are wanting to express in the schedule configurations.  We plan to update this section to be driven by `configuration` versus `code`.  That will make it much easier for the admin to change without the need to update the code.

Thanks!

--
jsynack

**Reply by NevilleGit** [reply]
Hi @jsynack ,

Thanks for offering the answer, it really helps!

I'd like to ask that where or what script can I check if I encounter that the Virtual Desktop instances are not following the custom scheduler (such as I set 8 am - 9 am, but it doesn't stop at 9 am. )

Where can I check the start/stop rule or scheduler rule?

Thank you.

**Reply by mcrozes** [reply]
Hello @NevilleGit  -

The script responsible for this is available here https://github.com/awslabs/scale-out-computing-on-aws/blob/main/source/soca/cluster_manager/web_interface/scheduled_tasks/manage_dcv_instances_lifecycle.py . Please note we have reworked this logic entirely and will publish an updated version soon.

`(such as I set 8 am - 9 am, but it doesn't stop at 9 am. )`

Please note a desktop will be stopped only if inactive, meaning:
- Should not have any active DCV connection for the past hours  https://github.com/awslabs/scale-out-computing-on-aws/blob/main/source/soca/cluster_manager/web_interface/config.py#L121 &  https://github.com/awslabs/scale-out-computing-on-aws/blob/main/source/soca/cluster_manager/web_interface/config.py#L140 
- CPU usage should be below a threshold (15% by default) https://github.com/awslabs/scale-out-computing-on-aws/blob/main/source/soca/cluster_manager/web_interface/config.py#L113

**Reply by NevilleGit** [reply]
hi @mcrozes ,

I've checked the inactive meaning you provided, so i launched a new, clean linux desktop session and set the scheduler [10:30-11:30], then I stop the desktop session at 9:30 to see if it follows the scheduler.

At 10:39, it started, I did not connect or do any operation while it is running, but it still remains running state after 11:30 (and now is 12:11).

i also check the CloudTrail for "StopInstances", and there is no any trail record shows.
 
is there any log that I can check for this situation?  thank you.
